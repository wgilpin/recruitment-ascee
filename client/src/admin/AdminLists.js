import React from 'react';
import Loader from 'react-loader-spinner';
import TableStyles from '../evidence/TableStyles';
import Styles from '../common/Styles';
import RoundImage from '../common/RoundImage';
import FabButton from '../common/fabButton';
import deleteImg from '../images/delete-white.svg';
import FetchData from '../common/FetchData';
import FindItem from './FindItem';
import Confirm from '../common/Confirm';

const propTypes = {};

const defaultProps = {};

const styles = {
  ...TableStyles.styles,
  ...Styles.styles,
  outer: {
    maxWidth: '400px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  listbox: {
    display: 'table',
    width: '300px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  li: {
    height: '32px',
    borderStyle: 'solid',
    borderWidth: '1px',
  },
  textarea: {
    backgroundColor: Styles.themeColors.secondary,
    border: 'none',
    textAlign: 'left',
    padding: '8px',
    color: 'white',
  },
  input: {
    textAlign: 'left',
    width: '140px',
    height: '33px',
    padding: '4px',
    color: 'white',
    backgroundColor: Styles.themeColors.secondary,
    border: 'none',
  },
  centreDiv: {
    padding: '8px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  selectPrimary: {
    transform: 'scale(1.5)',
    color: 'white',
    backgroundColor: Styles.themeColors.primary,
    border: 'none',
    marginBottom: '24px',
  },
  fab: {
    float: 'right',
    marginTop: '30px',
    position: 'inherited',
  },
};

export default class AdminLists extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      lists: [],
      kind: 'character',
      selectedItem: {},
      showConfirm: false,
      loading: true,
    };
  }

  static list_kinds = {
    character: 'Red Character',
    inventory_type: 'Flagged Type',
    alliance: 'Red Alliance',
    corporation: 'Red Corporation',
    system: 'Red System',
    user: 'User Blocked From Applying',
    region: 'Region',
  };

  titleise(text) {
    return (text[0].toUpperCase() + text.slice(1)).replace(/_/g, ' ');
  }

  componentDidMount() {
    this.doLoad();
  }

  doFetch = (scope) => new FetchData(scope).get();
  doFetchDelete = (scope) => new FetchData(scope).delete();

  doLoad = () => {
    const lists = {};
    const promises = [];
    const fetcher = this.props.fetcher || this.doFetch;
    for (let kind of Object.keys(AdminLists.list_kinds)) {
      promises.push(
        fetcher({ id: kind, scope: 'admin/list' }).then(list => {
          lists[kind] = list.info;
        })
      );
    }
    Promise.all(promises).then(() => {
      const showInput =
        lists.length === 0 || lists[this.state.kind].length === 0;
      this.setState({ lists, showInput, loading: false });
    });
  };

  handleConfirmDelete(id) {
    const selectedItem = this.state.lists[this.state.kind].filter(
      it => it.id === id
    )[0];
    this.setState({
      showConfirm: true,
      selectedItem: selectedItem,
    });
  }

  handleDelete = async () => {
    const fetcher = this.props.fetcher || this.doFetchDelete;
    await fetcher({
      id: this.state.kind,
      scope: 'admin/list',
      param1: 'delete',
      param2: this.state.selectedItem.id,
    });
    this.setState({ showConfirm: false });
    this.componentDidMount();
  };

  handleChangeKind = event => {
    this.setState({
      kind: event.target.value,
      showInput: this.state.lists[event.target.value].length === 0,
      showAddMany: false,
    });
  };

  handleAddFab = () => {
    this.setState({ showInput: true });
  };

  handleDoAdd = async () => {
    await this.searchOne();
    this.setState({ showInput: false });
  };

  handleDoAddMany = () => {
    this.setState({ showAddMany: false });
  };

  makeListLine(item, idx) {
    const lineStyle = {
      ...(idx % 2 === 0 ? styles.isOdd : {}),
      ...styles.row,
    };
    const characterImgSrc = `https://image.eveonline.com/Character/${
      item.id
    }_64.jpg`;
    const typeImgSrc = `https://image.eveonline.com/Type/${item.id}_64.png`;
    return (
      <div key={item.id} style={lineStyle}>
        {this.state.kind === 'character' && (
          <div style={styles.cell}>
            <RoundImage src={characterImgSrc} />
          </div>
        )}
        {this.state.kind === 'inventory_type' && (
          <div style={styles.cell}>
            <img
              style={{ width: '20px', height: '20px' }}
              src={typeImgSrc}
              alt=" "
            />
          </div>
        )}

        <div style={{ ...styles.cell, verticalAlign: 'middle' }}>
          {item.name}
        </div>
        <div
          style={{ ...styles.cell, verticalAlign: 'middle', cursor: 'pointer' }}
        >
          <img
            src={deleteImg}
            onClick={() => this.handleConfirmDelete(item.id)}
            alt="delete"
          />
        </div>
      </div>
    );
  }

  render() {
    if (this.state.loading) {
      return <Loader type="Puff" color="#01799A" height="100" width="100" />;
    }
    const { kind, lists } = this.state;
    return (
      <div style={styles.outer}>
        <h2>Redlists</h2>
        <select style={styles.selectPrimary} onChange={this.handleChangeKind}>
          {Object.entries(AdminLists.list_kinds).map(
            ([option_kind, option_text]) => (
              <option value={option_kind}>{option_text}</option>
            )
          )}
        </select>
        <div style={styles.listbox}>
          {(lists[kind] || [])
            .sort((a, b) => (a.name < b.name ? -1 : ( a.name > b.name ? 1 : 0)))
            .map((item, idx) => this.makeListLine(item, idx))}
        </div>
        {this.state.showInput && (
          <FindItem kind={this.state.kind} onChange={this.doLoad} />
        )}
        {!this.state.showInput && (
          <FabButton
            onClick={this.handleAddFab}
            icon="add"
            color="#c00"
            size="40px"
            style={styles.fab}
          />
        )}
        {this.state.showConfirm && (
          <Confirm
            text={`Remove item from redlist`}
            onConfirm={this.handleDelete}
            onClose={() => this.setState({ showConfirm: false })}
          />
        )}
      </div>
    );
  }
}

AdminLists.propTypes = propTypes;
AdminLists.defaultProps = defaultProps;
