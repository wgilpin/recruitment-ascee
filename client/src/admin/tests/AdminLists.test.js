import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import AdminLists from './../AdminLists';

let mockFetch;

beforeEach(() => {
  const serverState = {
    'character': [{"id":299590276,"name":"Major Sniper"},{"id":91027948,"name":"Recruiter TS"},{"id":94476783,"name":"Miyogi San"}],
    'inventory_type': [{"id":40520,"name":"Large Skill Injector"},{"id":11568,"name":"Avatar Blueprint"},{"id":40519,"name":"Skill Extractor"},{"id":23773,"name":"Ragnarok"},{"id":11567,"name":"Avatar"},{"id":23774,"name":"Ragnarok Blueprint"}],
    'alliance': [{id: 123, name: 'alliance1'}],
    'corporation': [{id: 1234, name: 'corp1'}],
    'system': [{"id":30004759,"name":"1DQ1-A"}],
    'user': [{"id":299590276,"name":"Major Sniper"}],
  };

  mockFetch = jest.fn().mockImplementation((scope, params) => {
    
    if (serverState[scope.id]) {
      return Promise.resolve({
        info: serverState[scope.id],
      });
    } else {
      console.error("Scope Not Valid"+scope.id);
      throw new Error("Scope Not Valid");
    }
  });
});

describe('<AdminRoles>', () => {
  it('renders without crashing', () => {
    shallow(<AdminLists fetcher={mockFetch} />);
  });

  it('matches snapshot', async () => {
    const tree = renderer.create(<AdminLists fetcher={mockFetch} />);
    await Promise.resolve();
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
