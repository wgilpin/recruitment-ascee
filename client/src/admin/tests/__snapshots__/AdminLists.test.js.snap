// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminRoles> matches snapshot 1`] = `
<div
  style={
    Object {
      "marginLeft": "auto",
      "marginRight": "auto",
      "maxWidth": "400px",
    }
  }
>
  <h2>
    Redlists
  </h2>
  <select
    onChange={[Function]}
    style={
      Object {
        "backgroundColor": "#01799A",
        "border": "none",
        "color": "white",
        "marginBottom": "24px",
        "transform": "scale(1.5)",
      }
    }
  >
    <option
      value="character"
    >
      Red Character
    </option>
    <option
      value="inventory_type"
    >
      Flagged Type
    </option>
    <option
      value="alliance"
    >
      Red Alliance
    </option>
    <option
      value="corporation"
    >
      Red Corporation
    </option>
    <option
      value="system"
    >
      Red System
    </option>
    <option
      value="user"
    >
      User Blocked From Applying
    </option>
  </select>
  <div
    style={
      Object {
        "display": "table",
        "marginLeft": "auto",
        "marginRight": "auto",
        "width": "300px",
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#151515",
          "display": "table-row",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
          }
        }
      >
        <img
          alt="Avatar"
          onClick={[Function]}
          src="https://image.eveonline.com/Character/299590276_64.jpg"
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": "50%",
              "boxShadow": "1px 3px 1px #111",
              "height": 32,
              "width": 32,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        Major Sniper
      </div>
      <div
        style={
          Object {
            "cursor": "pointer",
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        <img
          alt="delete"
          onClick={[Function]}
          src="delete-white.svg"
        />
      </div>
    </div>
    <div
      style={
        Object {
          "display": "table-row",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
          }
        }
      >
        <img
          alt="Avatar"
          onClick={[Function]}
          src="https://image.eveonline.com/Character/94476783_64.jpg"
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": "50%",
              "boxShadow": "1px 3px 1px #111",
              "height": 32,
              "width": 32,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        Miyogi San
      </div>
      <div
        style={
          Object {
            "cursor": "pointer",
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        <img
          alt="delete"
          onClick={[Function]}
          src="delete-white.svg"
        />
      </div>
    </div>
    <div
      style={
        Object {
          "backgroundColor": "#151515",
          "display": "table-row",
        }
      }
    >
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
          }
        }
      >
        <img
          alt="Avatar"
          onClick={[Function]}
          src="https://image.eveonline.com/Character/91027948_64.jpg"
          style={
            Object {
              "backgroundColor": "black",
              "borderRadius": "50%",
              "boxShadow": "1px 3px 1px #111",
              "height": 32,
              "width": 32,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        Recruiter TS
      </div>
      <div
        style={
          Object {
            "cursor": "pointer",
            "display": "table-cell",
            "paddingLeft": "6px",
            "textAlign": "left",
            "verticalAlign": "middle",
          }
        }
      >
        <img
          alt="delete"
          onClick={[Function]}
          src="delete-white.svg"
        />
      </div>
    </div>
  </div>
  <img
    alt="Add Alt"
    onClick={[Function]}
    src="add.gif"
    style={
      Object {
        "backgroundColor": "#c00",
        "borderRadius": "50%",
        "bottom": "30px",
        "boxShadow": "1px 3px 1px #111",
        "float": "right",
        "height": 32,
        "marginTop": "30px",
        "position": "inherited",
        "right": "30px",
        "width": 32,
      }
    }
  />
</div>
`;
